File: old-python/loc.py
from typing import *
from dataclasses import dataclass

@dataclass
class Loc:
    path: str
    line: int = 1
    col: int = 1
    char: int = 0

    def advance(self, code: str) -> Self:
        if code[self.char] == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        self.char += 1
        return self

    def copy(self) -> "Loc":
        return Loc(self.path, self.line, self.col, self.char)

    def __repr__(self):
        return f"{self.path}:{self.line}:{self.col}"

File: src/loc.rs
use std::fmt::{Debug, Display, Formatter};

#[derive(Clone, PartialEq)]
pub struct Loc {
    path: Box<String>,
    line: usize,
    col: usize,
    char: usize,
}

impl Debug for Loc {
    fn fmt(&self, fmt: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
        write!(fmt, "{self}")
    }
}

impl Display for Loc {
    fn fmt(&self, fmt: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
        write!(fmt, "{}:{}:{}", self.path, self.line, self.col)
    }
}

impl Loc {
    pub fn new(path: &String) -> Self {
        Self {
            path: Box::new(path.clone()),
            line: 1,
            col: 1,
            char: 0,
        }
    }

    // FIXME: line reporting is off on the first character of a line
    pub fn advance(&mut self, ch: char) -> char {
        self.char += 1;
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        ch
    }
}
